File: sandbox/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/application/exception/InvalidPromotionException.java
package com.simple.ecommerce.application.exception;

/**
 * Exception thrown when an invalid promotion is applied or when there's an issue with a promotion.
 */
public class InvalidPromotionException extends RuntimeException {
    
    public InvalidPromotionException(String message) {
        super(message);
    }
    
    public InvalidPromotionException(String message, Throwable cause) {
        super(message, cause);
    }
    
    public InvalidPromotionException(String promotionCode, String reason) {
        super("Invalid promotion code '" + promotionCode + "': " + reason);
    }
} 

File: Projects/Testtrack/p911 - Simple eCommerce backend/application/exception/invalid_promotion_exception.py
from typing import Optional


class InvalidPromotionException(Exception):
    def __init__(self, message_or_code: str, reason_or_cause: Optional[str] = None,
                 cause: Optional[Exception] = None):
        if reason_or_cause is not None and isinstance(reason_or_cause, str):
            message = f"Invalid promotion code '{message_or_code}': {reason_or_cause}"
            super().__init__(message)
            self.__cause__ = cause
        else:
            super().__init__(message_or_code)
            self.__cause__ = reason_or_cause if isinstance(
                reason_or_cause, Exception) else cause
